### PR TITLE
Replace hand-written Makefiles with CMake

### PR DESCRIPTION
**Pull request by davidebeatrici**
SoftEther VPN Patch Acceptance Policy:
http://www.softether.org/5-download/src/9.patch

I choose option 1.

### CODE REVIEW

**Comment by chipitsine** [comment]
it sounds like a good time to add osx to travis-ci matrix.

if you are not planning to do it yourself, I'll add it

**Comment by davidebeatrici** [comment]
Feel free to do it.

**Comment by chipitsine** [comment]
I would add cmake also to https://github.com/SoftEtherVPN/SoftEtherVPN/blob/master/debian/control#L5
(otherwise https://launchpad.net/~paskal-07/+archive/ubuntu/softethervpn will get broken)

**Comment by chipitsine** [comment]
also, it would be nice to have cross builds. it is easier to do with cmake, maybe next PR...

**Comment by davidebeatrici** [comment]
> I would add cmake also to https://github.com/SoftEtherVPN/SoftEtherVPN/blob/master/debian/control#L5
(otherwise https://launchpad.net/~paskal-07/+archive/ubuntu/softethervpn will get broken)

Done.

> also, it would be nice to have cross builds. it is easier to do with cmake, maybe next PR...

Indeed.

**Comment by davidebeatrici** [comment]
@chipitsine Pull request to add the Mac OS X target to Travis CI: #523

**Comment by chipitsine** [comment]
I do not agree with certain aspects. 

but let us choose your way, it is good to have osx builds

**Comment by davidebeatrici** [comment]
Feel free to explain what you don't like, feedbacks are very important.

**Comment by chipitsine** [comment]
currently. travis-ci caches openssl build which is located in travis-ci home. you seem have changed location to /usr/local, so cache will not work.

**Comment by davidebeatrici** [comment]
I didn't change the location.

OpenSSL in `/usr/local` is provided by Homebrew.

**Comment by chipitsine** [comment]
oh, sorry

**Comment by davidebeatrici** [comment]
No problem.

**Comment by davidebeatrici** [comment]
Ready to be reviewed.

**Comment by chipitsine** [comment]
let me explain my "approval":

code looks good. even if there's some issue, it is ok to break master for a while (and fix right after)

**Comment by davidebeatrici** [comment]
This pull request has been tested on:

- Debian 9.4
- FreeBSD 11.1
- OpenBSD 6.2

**Comment by chipitsine** [comment]
@moatazelmasry2 , can we merge this?

**Comment by moatazelmasry2** [comment]
I'll try to my best to get this reviewed today evening or tomorrow.

**Comment by norbusan** [comment]
I was just repackaging the current master for debian - updating the debian port, and when I was pointed here.

Would it be somehow possible to follow the FHS in *not* putting directories directly under `/usr`? I would suggest

```
/usr/lib/softether/vpnserver
/usr/lib/softether/vpnclient
/usr/lib/softether/vpnbridge
/usr/lib/softether/vpncmd
```

Or even better, make `/usr/lib/softether` a configure time option. If you want I can look into this and provide patches, too.

Thanks for consideration

**Comment by chipitsine** [comment]
there're many ways to improve cmake, for example implement cross builds.
or follow FHS.

but I'd leave them for separate PRs

**Comment by moatazelmasry2** [comment]
Yup. I also would love to have a build system that sucks less as soon as possible. And then start enhancing it

File: debian/control [hunk 1]
```diff
@@ -2,7 +2,7 @@ Source: softether-vpn
 Section: net
 Priority: optional
 Maintainer: Dmitry Orlov <me@mosquito.su>
-Build-Depends: debhelper (>= 7.0.50~), libncurses5-dev, linux-libc-dev, libssl-dev, zlib1g-dev, libreadline-dev, build-essential, dh-exec
+Build-Depends: debhelper (>= 7.0.50~), libncurses5-dev, linux-libc-dev, libssl-dev, zlib1g-dev, libreadline-dev, build-essential, cmake, dh-exec
```
Comment (moatazelmasry2) [inline debian/control]: Why do you need cmake for runtime?

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+
+add_definitions(-D_REENTRANT -DREENTRANT -D_THREAD_SAFE -D_THREADSAFE -DTHREAD_SAFE -DTHREADSAFE -D_FILE_OFFSET_BITS=64)
+
+set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -fsigned-char")
+
+# Add /src to the include paths
+include_directories(.)
+
+if(WIN32)
+  message(FATAL_ERROR "Windows compilation via CMake is currently not supported.")
```
Comment (moatazelmasry2) [inline src/CMakeLists.txt]: Should't the program exit here?

Comment (davidebeatrici) [reply src/CMakeLists.txt]: From [CMake's documentation](https://cmake.org/cmake/help/v3.0/command/message.html):
`FATAL_ERROR    = CMake Error, stop processing and generation`

**Comment by moatazelmasry2** [comment]
Many thanks for the exploitation. Your PR has been merged

**Comment by davidebeatrici** [comment]
You're welcome.

**Pull request merged**